### PR TITLE
Add execution status to PipelineExecutor.

### LINKED ISSUE #63979: LOGICAL_ERROR Pipeline for PushingPipelineExecutor was finished before all data was inserted
I'm able to reproduce smth similar with the following:

``` sh
diff --git a/src/Server/TCPHandler.cpp b/src/Server/TCPHandler.cpp
index ae2f150c4a1..d0911804063 100644
--- a/src/Server/TCPHandler.cpp
+++ b/src/Server/TCPHandler.cpp
@@ -910,6 +910,8 @@ void TCPHandler::processInsertQuery()

     auto run_executor = [&](auto & executor, Block processed_data)
     {
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+
         /// Made above the rest of the lines,
         /// so that in case of `start` function throws an exception,
         /// client receive exception before sending data.
```

``` sh
❯ cat repro.sql
drop table if exists mv_table_1;
drop table if exists mv_table;
drop table if exists null_table;

set max_execution_time = 0.1, timeout_overflow_mode = 'break';

CREATE TABLE null_table (str String) ENGINE = Null;
CREATE MATERIALIZED VIEW mv_table (str String) ENGINE = MergeTree ORDER BY str AS SELECT str AS str FROM null_table;
CREATE MATERIALIZED VIEW mv_table_1 (str Decimal(38, 7)) ENGINE = MergeTree ORDER BY str AS SELECT str AS str FROM null_table;

INSERT INTO null_table VALUES ('test');
```

```
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Pipeline for PushingPipelineExecutor was finished before all data was inserted. (LOGICAL_ERROR)
```

---

Stepped on this while trying to reproduce https://s3.amazonaws.com/clickhouse-test-reports/63861/af84e3e1a9c8c3b62f738f231e457efc8ea301ea/ast_fuzzer__tsan_/fatal.log

Not sure if it is connected anyhow with https://github.com/ClickHouse/ClickHouse/issues/52234 or not. Probably yes, since it requires `max_execution_time = 0.1, timeout_overflow_mode = 'break'`.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add execution status to PipelineExecutor. Avoid LOGICAL_ERROR if the pushing pipeline was canceled by timeout.
Replace `Pipeline for PushingPipelineExecutor was finished before all data was inserted. (LOGICAL_ERROR)` to `QUERY_WAS_CANCELLED`. Closes #63979

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 04286bc270f9f473a07bc4ae27ae61d96256f775 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68291/04286bc270f9f473a07bc4ae27ae61d96256f775/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/99282e526a8aeb175e9f3e69fc9385070d03798a/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Processors/Executors/PipelineExecutor.cpp [hunk 1]
```diff
@@ -289,11 +295,13 @@ void PipelineExecutor::executeStepImpl(size_t thread_num, std::atomic_bool * yie
                 Queue async_queue;
 
                 /// Prepare processor after execution.
-                if (!graph->updateNode(context.getProcessorID(), queue, async_queue))
-                    cancel();
+                auto status = graph->updateNode(context.getProcessorID(), queue, async_queue);
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/Executors/PipelineExecutor.cpp:298]: If the returned status is Cancelled, there is no need to call `cancel` again cause it was already done.
Previously, we could not separate `Cancel` and `Exception` indeed and just called `cancel` twice. This was not a bug, but I would better avoid calling `cancel` the second time.

File: src/Processors/Executors/PipelineExecutor.h [hunk 2]
```diff
@@ -48,8 +48,20 @@ class PipelineExecutor
 
     const Processors & getProcessors() const;
 
+    enum class ExecutionStatus
+    {
+        NotStarted,
+        Executing,
+        Finished,
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/Executors/PipelineExecutor.h:55]: Yes,
but why not have this information? :)

File: src/Processors/Executors/PipelineExecutor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Processors/Executors/PipelineExecutor.cpp:298]: ok

**Review by nickitat — APPROVED** [review]

**Pull request merged**